### PR TITLE
Use common methods between interceptors

### PR DESCRIPTION
**Pull request by emcfarlane**
`streamingState` renamed to `state` and now shared between all interceptor wrappers. Allows for common code to be reused easily and top level logic to be simple.
I've added documentation for which methods are safe to call concurrently and tried to remove some redundant locking.

Changes in behaviour:
- On error unary flows reported a 0 message size. Now message size attribute is removed, no size is reported, same as the streaming case.

Benchmark changes, compared to main:
```
goos: darwin
goarch: arm64
pkg: connectrpc.com/otelconnect
                           │   base.txt   │               new.txt                │
                           │    sec/op    │    sec/op     vs base                │
StreamingWithInterceptor-8   47.87µ ± ∞ ¹   47.10µ ± ∞ ¹       ~ (p=1.000 n=1) ²
UnaryWithInterceptor-8       31.55µ ± ∞ ¹   31.52µ ± ∞ ¹       ~ (p=1.000 n=1) ²
geomean                      38.86µ         38.53µ        -0.85%
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05

                           │   base.txt    │                new.txt                 │
                           │     B/op      │     B/op       vs base                 │
StreamingWithInterceptor-8   45.14Ki ± ∞ ¹   38.76Ki ± ∞ ¹        ~ (p=1.000 n=1) ²
UnaryWithInterceptor-8       38.50Ki ± ∞ ¹   34.05Ki ± ∞ ¹        ~ (p=1.000 n=1) ²
geomean                      41.69Ki         36.33Ki        -12.85%
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05

                           │  base.txt   │               new.txt                │
                           │  allocs/op  │  allocs/op   vs base                 │
StreamingWithInterceptor-8   275.0 ± ∞ ¹   242.0 ± ∞ ¹        ~ (p=1.000 n=1) ²
UnaryWithInterceptor-8       246.0 ± ∞ ¹   222.0 ± ∞ ¹        ~ (p=1.000 n=1) ²
geomean                      260.1         231.8        -10.89%
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05
```
For performance I think next places to look at is changing the API for the attribute methods to an append style, removing the temporary slice allocations. Added `append` method to filter but will need to follow up with the rest in another PR.

### CODE REVIEW

File: interceptor.go [hunk 1]
```diff
@@ -78,245 +74,108 @@ func (i *Interceptor) getAndInitInstrument(isClient bool) *instruments {
...
 // WrapUnary implements otel tracing and metrics for unary handlers.
 func (i *Interceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
-	return func(ctx context.Context, request connect.AnyRequest) (connect.AnyResponse, error) {
-		requestStartTime := i.config.now()
-		req := &Request{
-			Spec:   request.Spec(),
-			Peer:   request.Peer(),
-			Header: request.Header(),
+	return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
+		startAt := i.config.now() // TODO: get from connect
```
Comment (emcfarlane) [inline interceptor.go:78]: Added TODO here, the only accurate way to calculate the time would be for connect to pass the start time.

File: interceptor.go [hunk 2]
```diff
@@ -78,245 +74,108 @@ func (i *Interceptor) getAndInitInstrument(isClient bool) *instruments {
...
+		request := &Request{
+			Spec:   req.Spec(),
+			Peer:   req.Peer(),
+			Header: req.Header(),
 		}
 		if i.config.filter != nil {
-			if !i.config.filter(ctx, req) {
-				return next(ctx, request)
+			if !i.config.filter(ctx, request) {
+				return next(ctx, req)
```
Comment (jhump) [inline interceptor.go:86]: Oof, these are kinda terrible names. Simply swapping `request` and `req` doesn't clarify anything IMO. Can we use more descriptive names to distinguish what these two things are?

File: payloadinterceptor.go [hunk 3]
```diff
@@ -15,42 +15,67 @@
...
+		}
+		return err
+	}
+	s.state.receive(s.ctx, msg)
+	return nil
 }
 
 func (s *streamingClientInterceptor) Send(msg any) error {
-	return s.send(msg, s.StreamingClientConn)
+	s.init()
```
Comment (jhump) [inline payloadinterceptor.go:60]: This could trivially result in deadlock: this calls `RequestHeader`, which _blocks_ until the request headers are received. However, the server may choose to not send headers until after it has received and processed data. So we must allow sending messages without blocking for request headers.

**Review by jhump — COMMENTED** [review]
I've not done a full review, but noticed something a little scary in the streaming client interceptor. I figured I'd leave feedback on this before reviewing everything else, in case that one comment requires any material refactoring.

File: payloadinterceptor.go (hunk 3 above)
Comment (jhump) [reply payloadinterceptor.go:60]: Shoot, never mind. Ignore this comment 🤦.

I was apparently seeing "ResponseHeader" even though this actually says "RequestHeader". My bad. Pay this no mind, and I'll continue reviewing.

File: interceptor.go (hunk 2 above)
Comment (emcfarlane) [reply interceptor.go:86]: Was trying to keep consistent of what `req` and `request` refer to but yep, will fix.

Comment (emcfarlane) [reply interceptor.go]: Switched to `call` var but refrained from renaming the type. It is request metadata derived from `connect.Request` objects. Call is better but maybe could be just be called info or metadata?

Comment (jhump) [reply interceptor.go]: > derived from `connect.Request` objects.

Except for client streams, which can have multiple requests 😛  (hence the name suggestion).

FWIW, I'm fine leaving the type name alone, since it's exported. But in that case, maybe `req` and `msg` would be more clear?

**Review by jhump — COMMENTED** [review]
The instruments init PR has been merged so this needs a rebase. In the future, it would be nicer to review the PRs sequentially if they actually have dependencies. If they don't actually have dependencies, let's not use stacked branches.

File: payloadinterceptor.go [hunk 4]
```diff
@@ -59,29 +84,47 @@ func (s *streamingClientInterceptor) CloseRequest() error {
...
-	shouldCall := s.requestClosed && !s.onCloseCalled
-	if shouldCall {
-		s.onCloseCalled = true
-	}
+	shouldClose := !wasClosed && s.requestClosed && s.responseClosed
 	s.mu.Unlock()
-	if shouldCall {
+	if shouldClose {
 		s.onClose()
 	}
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply payloadinterceptor.go:93]: Oh I thought storing the extra bool was weird 😅 . Would changing the var names help? `shouldClose == `!wasClose && hasClosed`. It's calling it when both request and response have been closed, from either side.

[the remark this answers is not in the document]
Comment (jhump) [reply payloadinterceptor.go:93]: > It's calling it when both request and response have been closed, from either side.

Hmm. So we intentionally call it 2x for every stream? What does the callback actually do? Hard to imagine it does something useful if it doesn't care whether it was the request or response side that was closed 🤔

[the remark this answers is not in the document]
Comment (emcfarlane) [reply payloadinterceptor.go:93]: I think it could just be `shouldClose = s.responseClosed` and an early return if already closed.

[the remark this answers is not in the document]
Comment (emcfarlane) [reply payloadinterceptor.go:93]: > 2x for every stream

Calling 1x when both `CloseRequest` and `CloseResponse` have been called.

File: interceptor.go [hunk 5]
```diff
@@ -78,245 +74,108 @@ func (i *Interceptor) getAndInitInstrument(isClient bool) *instruments {
...
-					requestKey,
-					conn.RequestHeader(),
-					i.config.requestHeaderKeys)...),
-			)
-			// inject the newly created span into the carrier
-			carrier := propagation.HeaderCarrier(conn.RequestHeader())
-			i.config.propagator.Inject(ctx, carrier)
-		}
+		// start is deferred until the first send or receive to ensure
+		// the request headers are available.
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply interceptor.go:137]: It's not to ensure the underlying http headers are available. It calls `RequestHeader()` on the stream to read the headers set by the caller, which can't be altered during the first `Send()`. I think there might be some issue with `Receive()` calling init tho, headers should wait until `Send()`.

File: interceptor.go [hunk 6]
```diff
@@ -78,245 +74,108 @@ func (i *Interceptor) getAndInitInstrument(isClient bool) *instruments {
...
+			state: state,
+		})
+		if err != nil {
+			state.setError(err)
 		}
-		span.SetAttributes(state.attributes...)
-		span.SetAttributes(headerAttributes(protocol, responseKey, conn.ResponseHeader(), i.config.responseHeaderKeys)...)
-		span.SetStatus(spanStatus(protocol, err))
-		instrumentation.duration.Record(ctx, i.config.now().Sub(requestStartTime).Milliseconds(), metric.WithAttributes(state.attributes...))
+		state.end(ctx, conn.ResponseHeader(), startAt)
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply interceptor.go:178]: Might leave for another PR. Currently can use the panic handler in connect-go to wrap in front. It's documented as the solution: https://github.com/bufbuild/connect-go/blob/70ab23fc231114a07c368d858051296f67fdbd29/recover.go#L23

File: interceptor.go (hunk 2 above)
Comment (emcfarlane) [reply interceptor.go]: I think this is resolve, but please flag if not. Use `call` for `*Request` and `req` `rsp` for connect's Request and Response.

File: interceptor.go [hunk 7]
```diff
@@ -78,245 +74,108 @@ func (i *Interceptor) getAndInitInstrument(isClient bool) *instruments {
...
+		}
+		return rsp, err
 	}
 }
 
 // WrapStreamingClient implements otel tracing and metrics for streaming connect clients.
 func (i *Interceptor) WrapStreamingClient(next connect.StreamingClientFunc) connect.StreamingClientFunc {
 	return func(ctx context.Context, spec connect.Spec) connect.StreamingClientConn {
-		requestStartTime := i.config.now()
+		startAt := i.config.now() // TODO: get from connect
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply interceptor.go]: Moved to `state.start` for all.

File: interceptor.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (emcfarlane) [reply interceptor.go]: `Receive()` no long calls `init`. There has to be a `Send` call beforehand.

File: interceptor_test.go [hunk 8]
```diff
@@ -1861,20 +1860,12 @@ func assertSpans(t *testing.T, want []wantSpans, got []trace.ReadOnlySpan) {
...
-		if span.Name() != want[i].spanName {
-			t.Errorf("span name not %s", want[i].spanName)
-		}
+		assert.False(t, span.EndTime().IsZero(), "span has not ended")
+		assert.Equal(t, want[i].spanName, span.Name(), "span name does not match")
 		gotEvents := span.Events()
-		if len(wantEvents) != len(gotEvents) {
-			t.Fatal("event lengths do not match")
-		}
+		require.Equal(t, len(wantEvents), len(gotEvents), "event lengths do not match")
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply interceptor_test.go]: Creating a PR for this, need to update the libraries.

File: payloadinterceptor.go [hunk 9]
```diff
@@ -15,42 +15,67 @@
...
+func (s *streamingClientInterceptor) init() {
+	s.requestStarted.Do(func() {
+		header := s.StreamingClientConn.RequestHeader()
+		s.ctx = s.state.start(s.ctx, header)
+	})
 }
 
 func (s *streamingClientInterceptor) Receive(msg any) error {
-	return s.receive(msg, s.StreamingClientConn)
+	s.init()
```
[the remark this answers is not in the document]
Comment (emcfarlane) [reply payloadinterceptor.go]: This was botched, thanks for catching. Removed the `init` call, I don't think it's needed at all either. Guaranteed to wait until a Send is called, and Send ensures `init` is.

[the remark this answers is not in the document]
Comment (emcfarlane) [reply payloadinterceptor.go]: `CloseRequest` will call it once `CloseResponse` has also been called, as together they call `onClose`.

File: payloadinterceptor.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (emcfarlane) [reply payloadinterceptor.go]: I think this is resolved, please unresolved if not.

File: interceptor_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (emcfarlane) [reply interceptor_test.go]: https://github.com/connectrpc/otelconnect-go/pull/145

File: payloadinterceptor.go [hunk 10]
```diff
@@ -23,34 +26,55 @@ import (
...
 	requestClosed  bool
 	responseClosed bool
-	onCloseCalled  bool
+}
+
+func (s *streamingClientInterceptor) init() {
+	s.requestStarted.Do(func() {
+		header := s.RequestHeader()
+		s.ctx = s.state.start(s.ctx, header)
+	})
```
Comment (jhump) [inline payloadinterceptor.go:42]: This seems problematic. We've already passed a context to the `next` handler in the interceptor chain. So adding a span to the context here is incorrect, right? How will trace/span info get propagated on-the-wire and to the server if we already passed along the wrong context to those operations?

File: payloadinterceptor.go [hunk 11]
```diff
@@ -23,34 +26,55 @@ import (
...
+	}
+	s.state.send(s.ctx, msg)
+	return nil
 }
 
 func (s *streamingClientInterceptor) CloseRequest() error {
+	s.init() // Ensure initialized even under error conditions.
 	err := s.StreamingClientConn.CloseRequest()
 	s.mu.Lock()
+	hasClosed := !s.requestClosed && s.responseClosed
```
Comment (jhump) [inline payloadinterceptor.go:74]: Is this even right? What does it mean to close the response _before_ closing the request? Does that work? What does it do? I would think that `CloseResponse` is what it means to actually **close** the stream, and if the request side hasn't been closed, it would be implicitly closed as part of closing the response.

Basically, "close request" really means **half**-close, and "close response" really means close. So waiting for both to be called feels potentially brittle.

File: payloadinterceptor.go [hunk 12]
```diff
@@ -59,29 +83,45 @@ func (s *streamingClientInterceptor) CloseRequest() error {
...
+	header := s.ResponseHeader()
+	s.state.end(s.ctx, header)
+}
+
 type streamingHandlerInterceptor struct {
 	connect.StreamingHandlerConn
 
-	receive func(any, connect.StreamingHandlerConn) error
-	send    func(any, connect.StreamingHandlerConn) error
+	ctx   context.Context //nolint:containedctx
```
Comment (jhump) [inline payloadinterceptor.go:103]: I think you might not need this (and thus not need the nolint directive) if you actually tracked the span instead of the ctx. So if `state` had a reference to the span, you wouldn't need to hang on to the right ctx here, right?

**Review by jhump — COMMENTED** [review]
I apologize that this took me so long to get through another pass. But this branch was a bear to review. I don't meant to be lecture-y or annoying, but it felt more like a 3k LOC than 300, and I have some thoughts on what made it such a challenge.

I think there are a few issues going on here:
1. Doing too many things in one PR. Ideally, perf and behavioral changes/fixes would be in a _separate_ PR from mechanical refactoring/code motion. Or, if you think it's important to merge them together, at least do them in separate commits and suggest the reviewer read the PR commit-by-commit.
   * There are some things which seem to be for performance (the most prominent thing in the PR description after all is the output of benchmarks). I think this includes the change to only retain filtered attributes, instead of filtering at the end.
    * There are some behavior changes/improvements, like trying to delay the start time for client streams to when the operation is actually initiated (when request headers are actually available).
    * There is a _lot_ of code motion. Some of this appears to be stylistic/aesthetic improvements to the existing state and streaming code. Some of this appears to be making it more DRY (like to make unary use the state abstraction).
    * Some of the code motion is subtle and hard to follow: there is a chunk of 80 lines in the old `WrapUnary` function that was moved to no fewer than five different places in `state.go`, and also considerably re-ordered.
2. Not providing a well-curated history. I also often prefer a single commit for a PR -- except if it makes the PR harder to follow. In that case, it is best to carefully curate the changes into commits in order to tell a story and make the diff easier for a reviewer to understand. Nearly everything here was in one commit (9ffb3196cd2a5fe0b41b6cdb82f312253845956e). But a lot of code motion is much more straight-forward when broken up into multiple steps, especially for cases where git/GitHub isn't good at rendering the diff. Here's an example sequence of commits that might have made it easier to follow:
   * Rename `streaming.go` to `state.go`.
   * Refactor the `state` type and its methods so there's less boiler-plate in callers. This would only impact existing callers (streaming stuff only).
   * Update the unary flow last in its own PR. This one changed the most and was the hardest to follow (though there's some confusing aspects to the client streaming conn, too). I think if you first broke up the existing block of 80 lines into more clearly delineated chunks with comments, then refactoring each chunk+comment into a call to `state` might have been easier to follow.

   I know there's a stigma associated with `git rebase` in some parts of Buf. But honestly, I _prefer_ rebase in most circumstances. One issue with the commit history is that it is tainted with commits from two unrelated, prior PRs. You have to get to commit 7 to even get to the start of this change. This is an issue with stacking PRs -- it is often best to have them reviewed serially, even if you implement them in parallel -- to avoid this issue. But if you're going to stack PRs, please rebase after the base branch is merged so that the commit history is easier to follow.

The above issues make it harder to really understand and provide good feedback, which also makes it considerably easier for correctness issues to sneak in. I've made a few comments that worry me regarding correctness. It could be that the code _is_ correct, but it's hard to tell because of so much subtle code motion.

When writing the code, it's totally fine to use a single big commit or lots of incoherent commits to get things working. But when you're done, particularly for complex refactoring PRs, it really helps to do a soft reset to `main` and then selective `git add` and `git add --patch` commands to curate the changes in each commit. This is also a good time to just _not_ add some of the changes, and maybe stash them or put them in separate branches for future PRs.

While I appreciate that you were trying to make the final shape of the code easier to read, just as important (or perhaps even more important) is the journey the code must take to get there. Also, there are some aspects where the old code was actually easier to read, even if it was less DRY -- like the handling of request vs. response/send vs. receive in the unary flow. That makes the value of this level of consolidation a bit questionable.

Honestly, I'm scared of having to do follow-up reviews as you try to reconcile my feedback. Some of the feedback is rather broad, which means addressing it may mean even more churn in the code, resulting in more hard-to-follow changes. I think it might be better to start this one at square one, at least so the commit history is easier to follow. You might even try rolling this out in several smaller PRs, each with increments of what you're trying to do instead of doing everything at once.
